python-lib/ontology_tagging/formatter_by_document.py

```diff
@@ -0,0 +1,442 @@
+            original_sentence (Span): the original sentence from the input dataframe
+            tags_in_document (List): List of tags found in the current document
+            keywords_in_document (List): List of keywords tags found in the current document
+            matched_sentences (List): List of matched sentences in the current document
+            language (AnyStr): Language of the current document
+
+        Returns:
+            Lists of tags, keywords and sentences enriched with 'sentence' matches
+
+        """
+        matches = self._matcher_dict[language](sentence, as_spans=True)
+        for match in matches:
+            keyword = match.text
+            tag = self._keyword_to_tag[language][
+                get_span_text(span=match, lemmatize=self.lemmatization)
+            ]
+            tags_in_document.append(tag)
+            keywords_in_document.append(keyword)
+            matched_sentences.append(original_sentence + "\n")
+        return tags_in_document, keywords_in_document, matched_sentences
+
+    def write_df_category(
+        self,
+        input_df: pd.DataFrame,
+        text_column: AnyStr,
+        language_column: AnyStr = None,
+    ) -> pd.DataFrame:
+        """Write the output dataframe for format 'one row per document (array columns)' when there are categories in the ontology
+
+        Returns:
+         pd.DataFrame: a dataframe with the following columns:
+             - all columns from the input dataframe input_df
+             - a column with lists of keywords
+             - one column by category, with lists of tags
+             - a column with concatenated matched sentences
+
+        """
+        start = perf_counter()
+        self._generate_columns_names(input_df)
+        input_df.progress_apply(
+            self._write_row_category, args=[False, language_column], axis=1
+        )
+        logging.info(
+            f"Tagging {len(input_df)} documents : Done in {perf_counter() - start:.2f} seconds."
+        )
+        return self._merge_df_columns_category(input_df, text_column)
+
+    def _write_row_category(
+        self, row: pd.Series, one_row_per_doc_json: bool, language_column: AnyStr = None
+    ) -> None:
+        """
+        Called by FormatterByDocument.write_df_category or FormatterByDocumentJson.write_df_category
+        Append the tag columns to the output dataframe self.output_df
+        If one_row_per_doc_json is True, new columns are:
+        - 'tag_json_categories' : dictionary of category (keys) and tags (value)
+        - 'tag_json_full': dictionary of category (keys).
+            Each category has for value a dictionary of tags and their associated datas (keywords, sentences, occurences of the tag)
+        If one_row_per_doc_json is False, new columns are :
+            - a column with lists of keywords
+            - one column by category, with lists of tags
+            - a column with concatenated matched sentences
```

Comment:
I think it would be more readable to separate the two implementations: one in  `FormatterByDocument` and one in `FormatterByDocumentJson`

Reply:
I did in #21, if you want to discuss it there 
 It was indeed a little bit blurry in one unique method. Thank you !